### ISSUE TITLE
Neo4j SET very long string

### ISSUE THREAD

**Issue by naisanza** [neo4j/neo4j#8701]
**Neo4j version:** [3.1.0 community Dockerfile](https://github.com/neo4j/docker-neo4j-publish/blob/36ac547bcffcce2c7c0f249b340d062ba08e2598/3.1.0/community/Dockerfile)
**Operating system:** Ubuntu 16.10 
**API/Driver:** [Python Kubernetes/Cypher](https://neo4j.com/developer/python/#neo4j-python-driver)

I'm trying to find out what I need to do to insert a very long string into a node property 

The length of the string is 251172 characters long

I'm unable to catch the error output from the debug log

```
...truncated...
zM2QxYTc1ZWJmMTY2MjY0Y2IwYzNkOTkzYzIyYjUxPiA8NTQzZDFhNzVlYmYxNjYyNjRjYjBjM2Q5OTNjMjJiNTE+IF0KL1ByZXYgMTgyMDQ5Nwo+PgolRW5kRXhpZlRvb2xVcGRhdGUgMTgyMjMxNwpzdGFydHhyZWYKMTgyNTUwOQolJUVPRgo=', doc.updated = timestamp() CREATE UNIQUE (apt) - [r:reference] -> (doc)"
                                                                                                                                                                                                                                                                                                                                                                                                                                        ^
INFO:neo4j.bolt:~~ [CLOSE]
```

**Comment by spacecowboy** [comment]
Hmm. Can you try using query parameter perhaps? You can see usage if you type `:help param` in the browser

**Comment by naisanza** [comment]
@spacecowboy I forgot to add that I'm using the [Neo4j Python Driver](https://neo4j.com/developer/python/#neo4j-python-driver)

Here's the neo4j `session.run()` section. Everything works except when I want to add a long super long string, at `app_file_base64`:

```
...truncated...
session.run(
    'MERGE (app:app {{name: "{}", type: "app"}}) '.format(app) +
    'MERGE (doc:{} {{name: "{}", type: "{}"}}) '.format(
        app_file_ext.lower()[1:],
        app_file,
        app_file_ext.lower()[1:]) +
    'ON CREATE SET app.name = "{}", app.type = "app", app.added = timestamp() '.format(
        app) +
    'ON CREATE SET doc.name = "{}", doc.type = "{}", doc.added = timestamp() '.format(
        app_file,
        app_file_ext.lower()[1:],
        app_file_sha256,
        'timestamp()') +
    'SET doc.size_b = "{}", doc.size_kb = "{}", doc.sha256 = "{}", doc.base64_len = {}, doc.updated = {} '.format(
        app_file_size,
        '{:.0f}'.format(app_file_size / 1024),
        app_file_sha256,
        app_file_base64, 
        'timestamp()') +
    'CREATE UNIQUE (app) - [r:reference] -> (doc)'

)
...truncated...
```

**Comment by spacecowboy** [comment]
@naisanza you can still use query paramters with the bolt drivers. See the manual for examples:
https://neo4j.com/docs/developer-manual/current/drivers/run-statements/#_auto_commit_transactions

Example:

```python
very_long_string = "..."
result = session.run("CREATE (person:Person {name: {mystring}})", 
                     {"mystring": very_long_string})
```

**Comment by Mats-SX** [comment]
For more on parameters, see the Cypher docs: https://neo4j.com/docs/developer-manual/current/cypher/syntax/parameters/

**Comment by naisanza** [comment]
@spacecowboy @Mats-SX okay that works! I'm wondering how come that works?

**Comment by Mats-SX** [comment]
It works because it is supposed to work :)

Why the original thing with the very long literal did _not_ work, however, I do not know. I ran some preliminary tests on the server, and it seems to be able to handle several hundred thousands characters on the query string input. There isn't any limit in the drivers either, at least by design.

Could you provide a log message or small standalone code snippet that reproduces the error?

I should mention that using parameters is recommended in either case, because it has several benefits with respect to query planning.

**Comment by naisanza** [comment]
@Mats-SX I just tested on two random 7mb and 12mb files, and it fails on the 7mb file, before it get's to the 12mb file. The 7mb.img file is successfully base64'd and sha256'd before reaching `session.run()`

`dd if=/dev/urandom of=7mb.img bs=512 count=14336`
`dd if=/dev/urandom of=12mb.img bs=512 count=24576`

Using query parameters work as usual:

![screenshot from 2017-02-07 03-25-29](https://cloud.githubusercontent.com/assets/1614425/22683256/20a87516-ece5-11e6-990e-76bdb5e92ce5.png)

**Comment by naisanza** [comment]
@Mats-SX Here's the full code

```
def database():
    """Create database entities for apps"""

    # Neo4j Configuration
    neo4j_server = '10.100.0.107'
    neo4j_server_port = '30120'
    neo4j_bolt_connection = 'bolt://{}:{}'.format(neo4j_server, neo4j_server_port)
    user = 'neo4j'
    password = 'neo4j'
    # user = input('User (default: neo4j): ') or 'neo4j'
    # password = input('Password for {}: '.format(user)) or 'neo4j'

    # Neo4j Database Connection
    driver = GraphDatabase.driver(neo4j_bolt_connection, auth=basic_auth(user, password))


    # Create Neo4j Nodes
    # wd = os.path.dirname(os.path.realpath(__file__)) + '/_ready'
    wd = r'/home/eric/neo4j-test'

    if os.path.exists(wd) is False:
        print('[!] path not found')
        exit(1)

    logging.info('[*] os.walk {}'.format(wd))
    for directory, subdirectories, files in os.walk(wd):

        if directory == os.path.dirname(os.path.realpath(__file__)) + '/_ready/APP' or True:
            for sub in subdirectories:

                logging.info('[*] os.walk {}/{}'.format(directory, sub))
                for app_directory, app_subdirectories, app_files in os.walk('{}/{}'.format(directory, sub)):

                    for app_file in app_files:

                        app_file_fullpath = '{}/{}/{}'.format(directory, sub, app_file)

                        app_file_stat = os.stat(app_file_fullpath)
                        app_file_size = app_file_stat.st_size

                        with open(app_file_fullpath, 'rb') as f:
                            print('[*] b64: {}'.format(app_file), end=' ')
                            app_file_base64 = base64.b64encode(f.read())

                            print(len(app_file_base64), app_file_base64[:64], '...')
                            # print(app_file_base64)

                        with open(app_file_fullpath, 'rb') as f:
                            print('[*] sha256: {}'.format(app_file), end=' ')
                            app_file_sha256 = hashlib.sha256()
                            app_file_sha256.update(f.read())
                            app_file_sha256.digest()
                            app_file_sha256 = app_file_sha256.hexdigest()

                            print(app_file_sha256, '...')

                        app_filename, app_file_ext = os.path.splitext(app_file)

                        app = sub

                        session = driver.session()
                        try:
                            session.run(
                                'MERGE (app:APP {{name: "{}", type: "APP"}}) '.format(app) +
                                'MERGE (doc:{} {{name: "{}", type: "{}"}}) '.format(
                                    app_file_ext.lower()[1:],
                                    app_file,
                                    app_file_ext.lower()[1:]) +
                                'ON CREATE SET app.name = "{}", app.type = "APP", app.added = timestamp() '.format(
                                    app) +
                                'ON CREATE SET doc.name = "{}", doc.type = "{}", doc.added = timestamp() '.format(
                                    app_file,
                                    app_file_ext.lower()[1:],
                                    app_file_sha256,
                                    'timestamp()') +
                                'SET doc.size_b = "{sizeb}", doc.size_kb = "{sizekb}", doc.sha256 = "{sha}", doc.base64 = {{app_file_base64}}, doc.updated = {time} '.format(
                                    sizeb=app_file_size,
                                    sizekb='{:.0f}'.format(app_file_size / 1024),
                                    sha=app_file_sha256,
                                    time='timestamp()') +
                                'CREATE UNIQUE (app) - [r:reference] -> (doc)',
                                {"app_file_base64": app_file_base64}
                            )

                            """Testing for literal bug"""

                            # session.run(
                            #     'MERGE (app:APP {{name: "{}", type: "APP"}}) '.format(app) +
                            #     'MERGE (doc:{} {{name: "{}", type: "{}"}}) '.format(
                            #         app_file_ext.lower()[1:],
                            #         app_file,
                            #         app_file_ext.lower()[1:]) +
                            #     'ON CREATE SET app.name = "{}", app.type = "APP", app.added = timestamp() '.format(
                            #         app) +
                            #     'ON CREATE SET doc.name = "{}", doc.type = "{}", doc.added = timestamp() '.format(
                            #         app_file,
                            #         app_file_ext.lower()[1:],
                            #         app_file_sha256,
                            #         'timestamp()') +
                            #     'SET doc.size_b = "{sizeb}", doc.size_kb = "{sizekb}", doc.sha256 = "{sha}", doc.base64 = {app_file_base64}, doc.updated = {time} '.format(
                            #         sizeb=app_file_size,
                            #         sizekb='{:.0f}'.format(app_file_size / 1024),
                            #         sha=app_file_sha256,
                            #         app_file_base64=app_file_base64,
                            #         time='timestamp()') +
                            #     'CREATE UNIQUE (app) - [r:reference] -> (doc)'
                            # )
                        except:
                            raise

                        session.close()

                        pass

                # session = driver.session()
                # session.run('create (n:APP {})')
                # session.close()
                pass

        # Add subdirectory as node.name
        # Add all files relationships to node.name
        pass

        #
        # session = driver.session()
        # session.run(
        #     "CREATE (a:Person {name: 'Arthur', title: 'King'})"
        # )
        # session.close()
```

**Comment by Mats-SX** [comment]
@naisanza Thank you. We'll take a look at it.

**Comment by sherfert** [comment]
Hey @naisanza (and @Mats-SX )
I am not sure anyone has followed up on that. Could you provide a more self-contained version of that script? It would be awesome if you could include all necessary imports and remove all parts that are not 100% needed to trigger the bug. Thanks!

**Comment by naisanza** [comment]
@sherfert `app_file_fullpath` is a file created by `dd if=/dev/urandom of=7mb.img bs=512 count=14336` and `dd if=/dev/urandom of=12mb.img bs=512 count=24576`

And it's just called inside the function

**Comment by zhenlineo** [comment]
This issue is related to https://github.com/neo4j/neo4j-python-driver/issues/201 and https://github.com/neo4j/neo4j-python-driver/issues/165

The fix is https://github.com/neo4j/neo4j-python-driver/pull/206 and the fix will come in a 1.5 patch release soon.

**Comment by zhenlineo** [comment]
Hi @naisanza,

The bug in python driver around handing big result has been fixed in 1.5.3 python driver release. Please let us know if there is any reminding issues with your queries. 

Zhen

**Comment by naisanza** [comment]
@zhenlineo thanks!

Using query parameters had solved it for me before, but if you say it has been fixed in the Python driver, then the above code should now work without using query parameters
